### DISCUSSION TITLE
Switch to drf-spectacular as OAS scheme generator

### DISCUSSION THREAD

**Discussion by n2ygk** [django-json-api/django-rest-framework-json-api#1082] [Ideas]
Per https://github.com/encode/django-rest-framework/discussions/8453 upstream DRF is deprecating the internal schema generator in lieu of recommending use of [drf-spectactular](https://drf-spectacular.readthedocs.io/en/latest/) so we'll need to do the same since our openapi schema support extends DRF's and there have been outstanding upstream improvements (such as https://github.com/encode/django-rest-framework/pull/7516) that will never be implemented in DRF.

**Comment by sliverc** [comment]
Thanks for bringing up this discussion.

So far schema generator has not been deprecated in DRF. Once it is, we should also follow and deprecate it in DJA. Till then, I am happy to integrate small improvements like doc updates or obvious errors etc.

As OpenAPI won't be a direct feature of DRF anymore and also not a requirement of the JSON:API spec I would love to see a separate library based on drf-spectacular to be developed like drf-spectacular-json-api. This would follow the spirit of [DRF](https://www.django-rest-framework.org/community/third-party-packages/) and [drf-yasg-json-api](https://github.com/glowka/drf-yasg-json-api) is a great example where this worked out well.

Would be anyone willing working on such a library?

**Reply by jokiefer** [reply]
Just some experience from my own project:

In my [project](https://github.com/mrmap-community/mrmap/blob/c6a3a35d3e521504d4ced81c80ab01d0e05fbce4/backend/extras/openapi.py) i implement a lot extensions based on the DRF JsonApi schema generator, to inject more information inside the generated openapi schema. (Don't rate the code quality... I know it is not the best.)

For example:
* the concrete sortable fields of a list endpoint
* the concrete filterable fields of a list endpoint (with datatypes)
* operation ids based on the model name to simplify a dynamic matching in the client
* adds descriptions to any field (full translation support for help text)
* adds supports support for some special formats like geojson
* ...

The reason why we need it is, that our frontend client dynamically initializes the [table component](https://github.com/mrmap-community/mrmap/blob/c6a3a35d3e521504d4ced81c80ab01d0e05fbce4/frontend/src/components/SchemaTable/index.tsx#L153)  based on the given openapi schema. For that it is needed that the sortable, filterable columns with there datatype are part of the openapi schema.

Our frontent client has also a [form component](https://github.com/mrmap-community/mrmap/blob/c6a3a35d3e521504d4ced81c80ab01d0e05fbce4/frontend/src/components/SchemaForm/index.tsx#L68), which initializes by the given schema. Translated help text for any field are injected from the openapi schema.

OpenApi is a great thing to tell the client everthing he need and it is to big to implement it inside this project. So it would be very nice if someone could start to work on a library.

**Reply by n2ygk** [reply]
+1 to that @jokiefer. It's what I've been doing as well and was disappointed that DRF is moving away from directly supporting this stuff given all the code I've contributed here and used in my own projects. However, I've also had a PR to DRF to make a small improvement to OAS schema support pending for several years and it's never going to be implemented, so we'll have to see if drf-spectacular can be used to do the same thing.

**Reply by sliverc** [reply]
I saw that drf-spectacular already includes support for a variety of different libraries. It actually would be best, if DJA support could be added to drf-spectacular directly. I do not use OpenAPI regularly, so would someone be open to talk to the drf-spectacular devs whether they would consider adding such support? That would be great.

**Reply by tfranzel** [reply]
Hey folks, I was just made aware of this discussion by @auvipy. I'm still new to this particular aspect of the ecosystem, but I think there 2 options available. Please have a look at https://github.com/tfranzel/drf-spectacular/issues/870#issuecomment-1325159767. I'm happy to assist you in this endeavor, if you wish to choose the latter option.

**Reply by sliverc** [reply]
@tfranzel Thanks for joining the discussion. Let me give you a little background on this. This is my opinion, but also backed up with decisions we have made in the past. Within Django REST framework JSON:API our main expertise and our main [goal](https://github.com/django-json-api/django-rest-framework-json-api#goals) are to conform with the [JSON:API spec](https://jsonapi.org/). Our second goal is to be as compatible with Django REST framework as possible.

So the reason we added OpenAPI support in the first place, was not because JSON:API specification requires it (1st goal) but rather to be as compatible as possible with Django REST framework (2nd goal). As Django REST framework is now deprecating this support, according to our 2nd goal, there is no need to support it anymore and I think we should also follow suite deprecating it. I also think that we are simply not the experts when it comes to OpenAPI support within this community.

However, some users still depend on OpenAPI support, so how do we go from here. I see two ways forward. One would be that a library drf-spectacular-json-api is created which adds Django REST framework JSON:API by devs who need this support. A second option would be to bake Django REST framework JSON:API support directly into drf-spectacular (with an optional dependency, of course).

As a maintainer of drf-spectacular would you be open to add optional Django REST framework JSON:API support (of course with the help of our community)? Or would you prefer that a library like drf-spectacular-json-api is created?

**Reply by tfranzel** [reply]
Thanks @sliverc for the clarification. I think the picture is slowly coming into focus. 

I tried reading up a bit about JSON:API and drf-yasg-json-api. Seems like they make a couple of changes, but it does not look like they are rewriting large parts of either library. We have an extensions system to bolt on stuff, but it looks like this needs more than that interface provides. So getting this to work with only spectacular's extensions will most likely not be enough. 

The thing is we have our own set of goals (on top of our readme) and I would prefer not make too much consessions in that regard. Assuming extensions will not suffice, I'm unsure if adding direct support would introduce too much noise on our end, but that is merely my gut feeling at the moment.

Doing a project like drf-yasg-json-api seems reasonable to me, since drf-yasg is structurally similar. I think we use a bit more functional programming for low level stuff, but the higher level stuff is similarily OO. Since this has worked before, it may very well work again.

At this point I would probably go the `drf-spectacular-json-api` route, given my still limited understanding of what is actually required.

**Reply by sliverc** [reply]
Thanks @tfranzel for your feedback and insight. I also think starting with a new library drf-spectacular-json-api is properly the best way forward. This way it can be experimented with how drf-spectecular and JSON:API best play together without affecting drf-spectecular directly. With this experience, drf-spectecular API might be improved. And I guess in the future, the decision whether JSON:API support can be added as an extension to drf-spectecular can always be reconsidered as the API improves.

The question remains then who would have the time to start such a library. I unfortunately as of time constraints but also not having experience and an actual use case with OpenAPI won't be able to assist directly. In case there needs to be changes directly in “Django REST framework JSON:API” I am very open to assist.

@n2ygk @jokiefer As you joined this discussion, would you be open to work on such a library? **To everyone** reading, is anyone interested helping out? I would love to see such a library but I know there needs to be an effort by the community.

**Reply by tfranzel** [reply]
awesome @sliverc! I'm happy to assist and answer questions too. Once there is a prototype, I can more easily gauge whether we are able to absorb it, if that is even desired at that point.

**Reply by jokiefer** [reply]
I could work on it, maybe at the end of the year. I can move/merge my own code base to this new lib. Maybe you could create a bear repo under the django-json-api community?

**Reply by sliverc** [reply]
@jokiefer Thanks. That would be great if you found time, and maybe there are others who can support you as well? In terms of a repo under the django-json-api community. For now, I see drf-spectecular-json-api as 3rdParty library, such as drf-spectecular is also a 3rdParty library for the REST framework and hosted in its own space. If it is in the django-json-api Github organization, I would see it as a 1stParty library where we as DJA maintainers are responsible maintaining it. I think in the state where it is not even created yet, we cannot commit to this. So I prefer for the group who wants to work on drf-spectecular-json-api to decide where to host it. Once it is well established we may consider it again, adding it directly to the DJA community.

**Reply by axieum** [reply]
I'm also happy to provide support should this third-party package go ahead. I think for now, users who have been using tfranzel/drf-spectacular are forced to move back to axnsan12/drf-yasg which does not support OpenAPI 3.0. 😭 Maybe we should add this to the documentation somewhere to prevent any confusion?

**Reply by hongfanmeng** [reply]
So, is it in development now?

**Reply by sliverc** [reply]
@axieum Thanks for offering your support. Just to clarify for now no-one is forced to move to OpenAPI 2.0 as OpenAPI 3.0 schema generation in DJA is still possible and we also still integrate small fixes just like the recently accepted PR #1124. Also, as far as I understand in DRF small schema generation fixes are also integrated for the time being. However, at some point it will be deprecated and removed and it will be good to get prepared for this. I assume @jokiefer hasn't had time yet to get started with `drf-spectecular-json-api`. However, if you are willing to get it started, I have a feeling people will follow and support you, someone simply needs to take the lead.

**Reply by jokiefer** [reply]
Hi folks!

i started the project here [drf-spectecular-json-api](https://github.com/jokiefer/drf-spectecular-json-api).
First testable version is available from [pypi](https://pypi.org/project/drf-spectacular-jsonapi/).

It contains not all possible describable schema feature for now. but the basics are done.

Happy testing.

**Reply by sliverc** [reply]
@jokiefer Awesome. While implementing, if you encounter any issues which need a change in DJA directly, let me know. Happy to assist.

**Reply by jokiefer** [reply]
@sliverc The only thing from DJA i needed to overwrite for now was the Pagination class and there `get_paginated_response_schema` function. But in my pov, this function shall return an openapi schema and this shall not be a part of the DJA project.

I think IF DRF drops openapi support this function would also be deprecated.
This was also the mention you discovered in [commet](https://github.com/encode/django-rest-framework/discussions/8453#discussioncomment-4208577) @tfranzel?

**Reply by tfranzel** [reply]
@jokiefer awesome stuff! Interesting to see the actual changes required. Like @sliverc said, feel free to also raise issues on my end. Happy to support you here.

Regarding your question. I don't think this function will be dropped. At the end of the day the interface will remain and you could see that as part of the interface. This is worth a discussion and the maintainers can certainly be swayed in that regard if we make a good case of it. We can come up with an alternate approach if necessary, but I don't think it will come to that. 

What was mentioned in that comment was more meant towards the required changes in the documentation not the interface itself.

**Reply by n2ygk** [reply]
I've finally made some time to take a look at https://github.com/jokiefer/drf-spectecular-json-api @jokiefer. I'll be testing it with https://github.com/columbia-it/django-jsonapi-training where I had previously done a very complete OAS implementation including a bunch of extensions I had to do for OAuth2. I will let you and @tfranzel know how it goes.

**Reply by n2ygk** [reply]
@jokiefer Do you happen to have a small repo that demonstrates using drf-spectacular-json-api? I've tried integrating it into my project and am getting weird problems with `./manage.py spectacular` or `generateschema` that I can't figure out. It seems to be getting the wrong `DEFAULT_SCHEMA_CLASS`. This might be a problem with those management commands.... When I force the schema class like this: 
```
./manage.py generateschema --generator_class drf_spectacular_jsonapi.schemas.openapi.JsonApiAutoSchema
```
It throws an error about an unexpected `url` parameter: 
```
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/rest_framework/management/commands/generateschema.py", line 36, in handle
    generator = generator_class(
                ^^^^^^^^^^^^^^^^
TypeError: ViewInspector.__init__() got an unexpected keyword argument 'url'
```

Similar when I use this:
```
./manage.py spectacular --generator-class drf_spectacular_jsonapi.schemas.openapi.JsonApiAutoSchema
...
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/drf_spectacular/management/commands/spectacular.py", line 57, in handle
    generator = generator_class(
                ^^^^^^^^^^^^^^^^
TypeError: ViewInspector.__init__() got an unexpected keyword argument 'urlconf'
```

**Reply by jokiefer** [reply]
@n2ygk sry. I don't have a smal demo repo. Could you pls publish your `django-jsonapi-training` as public repo? Otherwise i currently can't reproduce your issue.

**Reply by n2ygk** [reply]
@jokiefer give me a minute; It looks like I perhaps screwed up something else when experimenting with drf-spectacular.

**Reply by n2ygk** [reply]
@jokiefer OK so those errors had to do with junk laying around from some testing I was doing with drf-spectacular. Sorry about that.

However, I have a bunch of new errors that seems to possibly be triggered by relationship views I think because they have two parameters: `<pk>` and `<related_field>`.  See https://github.com/columbia-it/django-jsonapi-training/tree/spectacular:

tag `drf-spectacular` in that branch shows basic drf-spectacular operation. And the OAS schema document gets generated, in the default (non-JSONAPI) format.

tag `drf-spectacular-jsonapi-initial` is at the head of the branch and has the differences to add in `drfsj` after getting a good `drfs` setup working. (I had to comment out some DJA schema customization I had done.)

```
$ ./manage.py spectacular --file schema4.yml
...
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/drf_spectacular/generators.py", line 256, in parse
    operation = view.schema.get_operation(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/drf_spectacular_jsonapi/schemas/openapi.py", line 48, in get_operation
    return super().get_operation(path, path_regex, path_prefix, method, registry)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 96, in get_operation
    tags = self.get_tags()
           ^^^^^^^^^^^^^^^
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/drf_spectacular_jsonapi/schemas/openapi.py", line 143, in get_tags
    return [get_resource_name(context={"view": self.view})]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/rest_framework_json_api/utils.py", line 51, in get_resource_name
    resource_name = view.resource_name
                    ^^^^^^^^^^^^^^^^^^
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/rest_framework_json_api/views.py", line 422, in get_resource_name
    instance = getattr(self.get_object(), self.get_related_field_name())
                       ^^^^^^^^^^^^^^^^^
  File "/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/rest_framework/generics.py", line 92, in get_object
    assert lookup_url_kwarg in self.kwargs, (
AssertionError: Expected view CourseTermRelationshipView to be called with a URL keyword argument named "pk". Fix your URL conf, or set the `.lookup_field` attribute on the view correctly.

```

PS: For reference to how we implemented OAS in DJA, take a look at https://github.com/django-json-api/django-rest-framework-json-api/pull/772

Thanks for your help!

**Reply by n2ygk** [reply]
PS: RelationShipView subclasses GenericAPIView not ModelViewSet like the others...

**Reply by n2ygk** [reply]
Hey, awesome! I commented out the RelationShipViews and it generates a nice schema. I will open this as an issue in https://github.com/jokiefer/drf-spectecular-json-api. See tag drf-spectacular-jsonapi-no-relationshipviews

**Reply by n2ygk** [reply]
See https://github.com/jokiefer/drf-spectecular-json-api/discussions/20

**Reply by sliverc** [reply]
@n2ygk 
As issue https://github.com/jokiefer/drf-spectecular-json-api/discussions/20 has now been solved, do you think drf-spectecular-json-api is now on par with DJA?

If so, I would suggest we officially deprecate OAS in DJA as DRF did.

Three things I see we should do:
1. add a deprecation notice to the schema docs in the lines like [DRF](https://www.django-rest-framework.org/api-guide/schemas/)
2. add a deprecation warning when running the DJA schema command
3. add drf-spectecular-json-api to the list of our 3rdParty libraries.

What do you think?

**Reply by axieum** [reply]
This is great to see.

I support the above points, with the addition to fix the spelling in the repository name.

This rename shouldn't impact functionality as the package appears to have been published with the correct spelling.

**Reply by n2ygk** [reply]
I need to finish integrating it in my demo app to be sure, but it seems like it is better than the existing DRF schema stuff.

**Reply by sliverc** [reply]
@n2ygk Good. Let us know once you have it fully integrated and whether you find any other not covered edge cases.

@axieum True, the naming is indeed not consistent. Best report it in the upstream repository so it can be fixed.

**Reply by n2ygk** [reply]
@sliverc I think we are good to go with deprecating the old DRF schema stuff. I've got it working with my demo app and, after making a few teaks and documentation updates, I'll merge those [changes](https://github.com/columbia-it/django-jsonapi-training/pull/29).

@jokiefer has done a great job with drf-spectacular-json-api!
